src/core/sandbox.c

```diff
@@ -0,0 +1,801 @@
+		if (rc < 0)
+			goto fail;
+	}
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(ioctl), 50);
+	if (rc < 0)
+		goto fail;
+
+	/* kill */
+	rc = seccomp_rule_add(ctx, SCMP_ACT_ERRNO(EPERM), SCMP_SYS(kill), 1,
+		SCMP_A1(SCMP_CMP_EQ, SIGTSTP));
+	if (rc < 0)
+		goto fail;
+
+	rc = seccomp_rule_add(ctx, SCMP_ACT_ERRNO(EPERM), SCMP_SYS(kill), 1,
+		SCMP_A1(SCMP_CMP_EQ, SIGKILL));
+	if (rc < 0)
+		goto fail;
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(kill), 50);
+	if (rc < 0)
+		goto fail;
+
+	/* prctl */
+	rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(prctl), 4,
+		SCMP_A0(SCMP_CMP_EQ, PR_SET_NAME),
+		SCMP_A2(SCMP_CMP_EQ, 0),
+		SCMP_A3(SCMP_CMP_EQ, 0),
+		SCMP_A4(SCMP_CMP_EQ, 0));
+	if (rc < 0)
+		goto fail;
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(prctl), 50);
+	if (rc < 0)
+		goto fail;
+
+	/* fcntl */
+	const int fcntl_1arg_array[3] = {
+		F_GETFL,
+		F_GETFD,
+		F_SETLK
+	};
+	for (i = 0; i < 3; i++) {
+		rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(fcntl), 1,
+			SCMP_A1(SCMP_CMP_EQ, fcntl_1arg_array[i]));
+		if (rc < 0)
+			goto fail;
+	}
+
+	struct fcntl_2args_struct_t {
+		int cmd;
+		int arg;
+	};
+	const struct fcntl_2args_struct_t fcntl_2args_struct[3] = {
+		{ F_SETFD, FD_CLOEXEC          },
+		{ F_SETFL, O_RDONLY            },
+		{ F_SETFL, O_RDONLY|O_NONBLOCK }
+	};
+	for (i = 0; i < 3; i++) {
+		rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(fcntl), 2,
+			SCMP_A1(SCMP_CMP_EQ, fcntl_2args_struct[i].cmd),
+			SCMP_A2(SCMP_CMP_EQ, fcntl_2args_struct[i].arg));
+		if (rc < 0)
+			goto fail;
+	}
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(fcntl), 50);
+	if (rc < 0)
+		goto fail;
+
+	/* mprotect */
+	const int mprotect_prot_array[3] = {
+		PROT_NONE,
+		PROT_READ,
+		PROT_READ|PROT_WRITE
+	};
+	for (i = 0; i < 3; i++) {
+		rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(mprotect), 1,
+			SCMP_A2(SCMP_CMP_EQ, mprotect_prot_array[i]));
+		if (rc < 0)
+			goto fail;
+	}
+
+	rc = seccomp_syscall_priority(ctx, SCMP_SYS(mprotect), 60);
+	if (rc < 0)
+		goto fail;
+
+	/* mmap */
+	struct mmap_struct_t {
+		int prot;
+		int flags;
+	};
+	const struct mmap_struct_t mmap_anon_struct[4] = {
+		{ PROT_NONE,            MAP_PRIVATE|MAP_ANONYMOUS|MAP_NORESERVE },
+		{ PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS               },
+		{ PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_STACK     },
+		{ PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS     }
+	};
+	for (i = 0; i < 4; i++) {
+		rc = seccomp_rule_add(ctx, SCMP_ACT_ALLOW, SCMP_SYS(mmap), 4,
+			SCMP_A2(SCMP_CMP_EQ, mmap_anon_struct[i].prot),
+			SCMP_A3(SCMP_CMP_EQ, mmap_anon_struct[i].flags),
+			SCMP_A4(SCMP_CMP_EQ, 0xffffffff),
+			SCMP_A5(SCMP_CMP_EQ, 0));
+		if (rc < 0)
+			goto fail;
+	}
+
+	const struct mmap_struct_t mmap_noanon_struct[3] = {
+		{ PROT_READ,           MAP_PRIVATE               },
+		{ PROT_READ,           MAP_SHARED                },
+		{ PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE }
```

Comment:
There are many ways to make pages executable, and the only realistic way to protect from them in irssi (as oppose to with W^X/PaX MPROTECT or gcc flags) is to limit the files which open() can access. I don't believe this can be done easily with Seccomp, but it may be possible with unprivileged namespaces. As long as open() cannot take a file and mmap() it with executable pages, then allowing executable pages is not a problem. Things like ROP will always be a risk (at least until RAP becomes popular), but even with ROP, you cannot bypass seccomp protections.

However I agree with you, and I'd like to find some way to prevent adding arbitrary data as executable pages, but I can't think of any simple ways to do it with seccomp, other than waiting until all executable libraries are loaded before restricting mmap() which might be impractical. If there is no way to do this, then I guess I'll just add a comment mentioning it.